sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,419 @@
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = _pipeline.CreateRequest();
+
+            request.Method = HttpPipelineMethod.Get;
+
+            request.Headers.Add("secret", s_secret);
+
+            request.UriBuilder.Uri = s_endpoint;
+
+            request.UriBuilder.AppendQuery("api-version", AppServiceMsiApiVersion);
+
+            request.UriBuilder.AppendQuery("resource", Uri.EscapeDataString(resource));
+
+            if (!string.IsNullOrEmpty(clientId))
+            {
+                request.UriBuilder.AppendQuery("client_id", Uri.EscapeDataString(clientId));
+            }
+
+            return request;
+        }
+
+        private Request CreateCloudShellAuthRequest(string[] scopes, string clientId)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = _pipeline.CreateRequest();
+
+            request.Method = HttpPipelineMethod.Post;
+
+            request.Headers.Add(HttpHeader.Common.FormUrlEncodedContentType);
+
+            request.UriBuilder.Uri = s_endpoint;
+
+            request.Headers.Add("Metadata", "true");
+
+            var bodyStr = $"resource={Uri.EscapeDataString(resource)}";
+
+            if (!string.IsNullOrEmpty(clientId))
+            {
+                bodyStr += $"&client_id={Uri.EscapeDataString(clientId)}";
+            }
+
+            ReadOnlyMemory<byte> content = Encoding.UTF8.GetBytes(bodyStr).AsMemory();
+
+            request.Content = HttpPipelineRequestContent.Create(content);
+
+            return request;
+        }
+
+        private async Task<AccessToken> DeserializeAsync(Stream content, CancellationToken cancellationToken)
+        {
+            using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cancellationToken).ConfigureAwait(false))
+            {
+                return Deserialize(json.RootElement);
+            }
+        }
+
+        private AccessToken Deserialize(Stream content)
+        {
+            using (JsonDocument json = JsonDocument.Parse(content))
+            {
+                return Deserialize(json.RootElement);
+            }
+        }
+
+        private AccessToken Deserialize(JsonElement json)
+        {
+            string accessToken = null;
+
+            DateTimeOffset expiresOn = DateTimeOffset.MaxValue;
+
+            if (json.TryGetProperty("access_token", out JsonElement accessTokenProp))
+            {
+                accessToken = accessTokenProp.GetString();
+            }
+
+            if (json.TryGetProperty("expires_on", out JsonElement expiresOnProp))
+            {
+                // if s_msiType is AppService expires_on will be a string formatted datetimeoffset
+                if (s_msiType == MsiType.AppService)
+                {
+                    expiresOn = DateTimeOffset.Parse(expiresOnProp.GetString());
```

Comment:
What if this parse fails? Can this fail?

Reply:
I think all exceptional cases which happen during the flow of acquiring a token should use the new exception type I mentioned above. Issue #6712 tracks this.